File: src/main/java/com/smarthome/backend/server/api/modules/lg/LGController.java
package com.smarthome.backend.server.api.modules.lg;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.smarthome.backend.model.devices.DeviceTV;

/**
 * Controller für die Ausführung von Aktionen auf LG-TV-Geräten.
 * <p>
 * Verwaltet WebOSClient-Verbindungen zu mehreren LG-TVs und stellt alle Controls
 * (Media, TV, System, Application, Input, Source) zur Verfügung.
 * <p>
 * Basiert auf der Python-Implementierung von pywebostv.
 */
public class LGController {
    private static final Logger logger = LoggerFactory.getLogger(LGController.class);

    public static Boolean register(LGTV tv) {
        Boolean connected = false;
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("register() für LGTV (address={})", address);
        if (clientKey != null) {
            return true;
        }
        if (address == null || address.isBlank()) {
            logger.warn("register() abgebrochen: Adresse fehlt ");
            return false;
        }

        Path scriptPath = Paths.get("scripts", "pywebostv", "register.py").toAbsolutePath();
        ProcessBuilder processBuilder = new ProcessBuilder(
                "python",
                scriptPath.toString(),
                "CONNECT",
                "--ip",
                address);
        processBuilder.redirectErrorStream(true);

        Pattern keyPattern = Pattern.compile("'client_key'\\s*:\\s*'([^']+)'");
        try {
            Process process = processBuilder.start();
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.contains("Please accept the connect on the TV!")) {
                        logger.info("Pairing-Bestätigung am TV erforderlich ({})", address);
                    } else if (line.contains("Registration successful!")) {
                        logger.info("Registrierung erfolgreich ({})", address);
                    }
                    Matcher matcher = keyPattern.matcher(line);
                    if (matcher.find()) {
                        tv.setClientKey(matcher.group(1));
                        connected = true;
                    }
                }
            }

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                logger.error("PyWebOSTV-Skript beendet mit Exit-Code {} für {}", exitCode, address);
            }
        } catch (java.io.IOException e) {
            logger.error("Fehler beim Starten des PyWebOSTV-Skripts für {}", address, e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            logger.error("PyWebOSTV-Skript wurde unterbrochen für {}", address, e);
        }
        return connected;
    }

    public static void powerOn(LGTV tv) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        String macAddress = tv.getMacAddress();
        logger.info("powerOn() für LGTV (address={})", address);
        if (macAddress == null || macAddress.isBlank()) {
            logger.warn("powerOn() abgebrochen: MAC-Adresse fehlt ");
            return;
        }
        String params = "{\"mac\":\"" + macAddress + "\"}";
        callController(address, clientKey, "wol", params);
    }

    public static void screenOn(LGTV tv) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("screenOn() für LGTV (address={})", address);
        callController(address, clientKey, "system.screen_on", null);
    }

    public static void screenOff(LGTV tv) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("screenOff() für LGTV (address={})", address);
        callController(address, clientKey, "system.screen_off", null);
    }

    public static void setVolume(LGTV tv, int volume) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("setVolume() für LGTV (address={}, volume={})", address, volume);
        if (clientKey == null) {
            logger.warn("setVolume() abgebrochen: Client-Key fehlt ");
            return;
        }
        if (volume < 1 || volume > 100) {
            logger.warn("setVolume() abgebrochen: Volume außerhalb Bereich (1-100): {}", volume);
            return;
        }
        callController(address, clientKey, "media.set_volume", String.valueOf(volume));
    }

    public static Integer getVolume(LGTV tv) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("getVolume() für LGTV (address={})", address);
        if (clientKey == null) {
            logger.warn("getVolume() abgebrochen: Client-Key fehlt ");
            return null;
        }
        JsonObject response = callController(address, clientKey, "media.get_volume", null);
        if (response == null) {
            return null;
        }
        if (response.has("status") && "error".equalsIgnoreCase(response.get("status").getAsString())) {
            logger.warn("getVolume() fehlgeschlagen: {}", response);
            return null;
        }
        JsonElement resultElement = response.get("result");
        if (resultElement == null || !resultElement.isJsonObject()) {
            return null;
        }
        JsonObject resultObj = resultElement.getAsJsonObject();
        if (!resultObj.has("volumeStatus")) {
            return null;
        }
        JsonObject volumeStatus = resultObj.get("volumeStatus").getAsJsonObject();
        if (volumeStatus == null || !volumeStatus.has("volume")) {
            return null;
        }
        try {
            return volumeStatus.get("volume").getAsInt();
        } catch (Exception e) {
            logger.warn("getVolume() Volume ungültig: {}", resultObj);
            return null;
        }
    }

    public static void powerOff(LGTV tv) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("powerOff() für LGTV (address={})", address);
        callController(address, clientKey, "system.power_off", null);

    }

    public static void setChannel(LGTV tv, String channelId) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("setChannel() für LGTV (address={})", address);
        if (clientKey == null) {
            logger.warn("setChannel() abgebrochen: Client-Key fehlt ");
            return;
        }
        String escapedChannelId = channelId
                .replace("\\", "\\\\")
                .replace("\"", "\\\"");
        String params = "\"" + escapedChannelId + "\"";
        callController(address, clientKey, "tv.set_channel_with_id", params);
    }

    public static void startApp(LGTV tv, String appId) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("startApp() für LGTV (address={})", address);
        if (clientKey == null) {
            logger.warn("startApp() abgebrochen: Client-Key fehlt ");
            return;
        }
        String escapedChannelId = appId
                .replace("\\", "\\\\")
                .replace("\"", "\\\"");
        String params = "\"" + escapedChannelId + "\"";
        callController(address, clientKey, "application.launch", params);
    }

    public static void notify(LGTV tv, String message) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("notify() für LGTV (address={})", address);
        if (message == null) {
            logger.warn("notify() abgebrochen: Nachricht fehlt ");
            return;
        }
        String escapedMessage = message
                .replace("\\", "\\\\")
                .replace("\"", "\\\"");
        String params = "{\"message\":\"" + escapedMessage + "\"}";
        callController(address, clientKey, "system.notify", params);
    }

    public static List<DeviceTV.Channel> getChannels(LGTV tv) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("getChannels() für LGTV (address={})", address);
        if (clientKey == null) {
            logger.warn("getChannels() abgebrochen: Client-Key fehlt ");
            return null;
        }
        JsonObject response = callController(address, clientKey, "tv.channel_list", null);
        return parseChannels(response);
    }

    public static List<DeviceTV.App> getApps(LGTV tv) {
        String address = tv.getAddress();
        String clientKey = tv.getClientKey();
        logger.info("getApps() für LGTV (address={})", address);
        if (clientKey == null) {
            logger.warn("getApps() abgebrochen: Client-Key fehlt ");
            return null;
        }
        JsonObject response = callController(address, clientKey, "application.list_apps", null);
        return parseApps(response);
    }

    private static List<DeviceTV.Channel> parseChannels(JsonObject response) {
        if (response == null) {
            return null;
        }
        if (response.has("status") && "error".equalsIgnoreCase(response.get("status").getAsString())) {
            logger.warn("getChannels() fehlgeschlagen: {}", response);
            return null;
        }

        JsonObject resultElement = response.getAsJsonObject("result");
        if (resultElement == null) {
            return null;
        }

        JsonArray channelArray = resultElement.getAsJsonArray("channelList");
        if (channelArray == null) {
            return null;
        } 

        List<DeviceTV.Channel> channels = new ArrayList<>();
        for (JsonElement element : channelArray) {
            if (!element.isJsonObject()) {
                continue;
            }
            JsonObject channelObj = element.getAsJsonObject();
            String id = null;
            String name = null;
            Integer channelNumber = null;
            String channelType = null;
            Boolean hd = null;
            String imgUrl = null;
            if (channelObj.has("channelId")) {
                id = channelObj.get("channelId").getAsString();
            } 
            if (channelObj.has("majorNumber")) {
                channelNumber = channelObj.get("majorNumber").getAsInt();
            }
            if (channelObj.has("channelMode")) {
                channelType = channelObj.get("channelMode").getAsString();
            }
            if (channelObj.has("HDTV")) {
                hd = channelObj.get("HDTV").getAsBoolean();
            }
            if (channelObj.has("channelName")) {
                name = channelObj.get("channelName").getAsString();
            } 
            if (channelObj.has("imgUrl")) {
                imgUrl = channelObj.get("imgUrl").getAsString();
            }
            channels.add(new DeviceTV.Channel(id, name, channelNumber, channelNumber, channelType, hd, imgUrl));
        }
        return channels;
    }

    private static List<DeviceTV.App> parseApps(JsonObject response) {
        if (response == null) {
            return null;
        }
        if (response.has("status") && "error".equalsIgnoreCase(response.get("status").getAsString())) {
            logger.warn("getApps() fehlgeschlagen: {}", response);
            return null;
        }

        JsonArray applications = response.getAsJsonArray("result");
        if (applications == null) {
            return null;
        }

        List<DeviceTV.App> appsArray = new ArrayList<>();
        
        for(JsonElement element : applications) {
            if (!element.isJsonObject()) {
                continue;
            }
            JsonObject appObj = element.getAsJsonObject();
            String id = null;
            String name = null;
            String imgUrl = null;
            if (appObj.has("id")) {
                id = appObj.get("id").getAsString();
            } else {
                continue;
            }
            if (appObj.has("title")) {
                name = appObj.get("title").getAsString();
            } else {
                continue;
            }
            if (appObj.has("icon")) {
                imgUrl = appObj.get("icon").getAsString();
            }
            appsArray.add(new DeviceTV.App(id, name, imgUrl, appsArray.size() + 1));
        }
        
        return appsArray;
    }
    
    private static JsonObject callController(String address, String clientKey, String event, String params) {
        boolean isWol = "wol".equalsIgnoreCase(event);
        if (!isWol && clientKey == null) {
            logger.warn("PyWebOSTV-Aufruf abgebrochen: Client-Key fehlt ");
            return null;
        }
        if (address == null || address.isBlank()) {
            logger.warn("PyWebOSTV-Aufruf abgebrochen: Adresse fehlt ");
            return null;
        }
        Path scriptPath = Paths.get("scripts", "pywebostv", "controller.py").toAbsolutePath();
        List<String> command = new ArrayList<>();
        command.add("python");
        command.add(scriptPath.toString());
        command.add("--ip");
        command.add(address);
        if (clientKey != null) {
            command.add("--client-key");
            command.add(clientKey);
        }
        command.add("--event");
        command.add(event);
        if (params != null) {
            command.add("--params");
            command.add(escapeArgIfWindows(params));
        }

        ProcessBuilder processBuilder = new ProcessBuilder(command);
        processBuilder.redirectErrorStream(true);
        String lastJsonLine = null;
        try {
            Process process = processBuilder.start();

            try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    logger.info("PyWebOSTV-Skript output: {}", line);
                    String trimmed = line.trim();
                    if (trimmed.startsWith("{") && trimmed.endsWith("}")) {
                        lastJsonLine = trimmed;
                    }
                }
            }

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                logger.error("PyWebOSTV-Skript beendet mit Exit-Code {} für {}", exitCode, address);
            }
        } catch (java.io.IOException e) {
            logger.error("Fehler beim Starten des PyWebOSTV-Skripts für {}", address, e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            logger.error("PyWebOSTV-Skript wurde unterbrochen für {}", address, e);
        }
        if (lastJsonLine != null) {
            try {
                return JsonParser.parseString(lastJsonLine).getAsJsonObject();
            } catch (Exception e) {
                logger.warn("PyWebOSTV-Output ist kein gültiges JSON: {}", lastJsonLine, e);
            }
        }
        return null;
    }

    private static String escapeArgIfWindows(String value) {
        String osName = System.getProperty("os.name");
        if (osName != null && osName.toLowerCase().contains("win")) {
            return value.replace("\"", "\\\"");
        }
        return value;
    }

    public static String getSelectedApp(LGTV lgtv) {
        String address = lgtv.getAddress();
        String clientKey = lgtv.getClientKey();
        logger.info("getSelectedApp() für LGTV (address={})", address);
        if (clientKey == null) {
            logger.warn("getSelectedApp() abgebrochen: Client-Key fehlt ");
            return null;
        }
        JsonObject response = callController(address, clientKey, "application.get_current", null);
        if (response == null) {
            return null;
        }
        if (response.has("status") && "error".equalsIgnoreCase(response.get("status").getAsString())) {
            logger.warn("getSelectedApp() fehlgeschlagen: {}", response);
            return null;
        }
        JsonElement resultElement = response.get("result");
        if (resultElement == null) {
            return null;
        }
        return resultElement.getAsString();
    }

    public static String getSelectedChannel(LGTV lgtv) {
        String address = lgtv.getAddress();
        String clientKey = lgtv.getClientKey();
        logger.info("getSelectedChannel() für LGTV (address={})", address);
        if (clientKey == null) {
            logger.warn("getSelectedChannel() abgebrochen: Client-Key fehlt ");
            return null;
        }
        JsonObject response = callController(address, clientKey, "tv.get_current_channel", null);
        if (response == null) {
            return null;
        }
        if (response.has("status") && "error".equalsIgnoreCase(response.get("status").getAsString())) {
            logger.warn("getSelectedChannel() fehlgeschlagen: {}", response);
            return null;
        }
        JsonElement resultElement = response.get("result");
        if (resultElement == null) {
            return null;
        }
        if (resultElement.isJsonObject() && resultElement.getAsJsonObject().has("channelId")) {
            return resultElement.getAsJsonObject().get("channelId").getAsString();
        }
        return null;
    }

    public static Boolean getPower(LGTV lgtv) {
        return false;
    }

    public static Boolean getScreen(LGTV lgtv) {
        return false;
    }
}

File: src/main/node/com/smarthome/backend/server/api/modules/lg/lgController.ts
import { spawn } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { logger } from "../../../../logger.js";
import { App, Channel } from "../../../../model/devices/DeviceTV.js";
import type { LGTV } from "./lgtv.js";

type JsonValue = Record<string, unknown> | null;

export class LGController {
  static async register(tv: LGTV) {
    if (tv.clientKey) return true;
    if (!tv.address) return false;

    logger.info({ address: tv.address }, "register() fuer LGTV");
    const scriptPath = resolveScriptPath("scripts", "pywebostv", "register.py");
    const args = ["CONNECT", "--ip", tv.address];
    try{
      const { output, code } = await runPython(scriptPath, args);
    
      const keyMatch = output.match(/'client_key'\s*:\s*'([^']+)'/);
      if (keyMatch?.[1]) {
        tv.clientKey = keyMatch[1];
        tv.isConnected = true;
      }
      if (code !== 0) {
        logger.error(
          { code, output: output.trim() },
          "PyWebOSTV-Skript beendet mit Exit-Code fuer {}",
          tv.address
        );
      }
      return Boolean(tv.clientKey);
    } catch (err) {
      console.log("Fehler beim Starten des PyWebOSTV-Skripts für "+tv.address+": "+JSON.stringify(err));
      return false;
    }
  }

  static async powerOn(tv: LGTV) {
    logger.info({ address: tv.address }, "powerOn() fuer LGTV: "+tv.address);
    if (!tv.macAddress) return;
    const params = `{"mac":"${tv.macAddress}"}`;
    await this.callController(tv.address, tv.clientKey, "wol", params);
  }

  static async powerOff(tv: LGTV) {
    logger.info({ address: tv.address }, "powerOff() fuer LGTV: "+tv.address);
    await this.callController(tv.address, tv.clientKey, "system.power_off", null);
  }

  static async screenOn(tv: LGTV) {
    logger.info({ address: tv.address }, "screenOn() fuer LGTV");
    await this.callController(tv.address, tv.clientKey, "system.screen_on", null);
  }

  static async screenOff(tv: LGTV) {
    logger.info({ address: tv.address }, "screenOff() fuer LGTV");
    await this.callController(tv.address, tv.clientKey, "system.screen_off", null);
  }

  static async setVolume(tv: LGTV, volume: number) {
    logger.info({ address: tv.address, volume }, "setVolume() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("setVolume() abgebrochen: Client-Key fehlt");
      return;
    }
    if (volume < 1 || volume > 100) {
      logger.warn({ volume }, "setVolume() abgebrochen: Volume ausserhalb Bereich (1-100)");
      return;
    }
    await this.callController(tv.address, tv.clientKey, "media.set_volume", String(volume));
  }

  static async getVolume(tv: LGTV) {
    logger.info({ address: tv.address }, "getVolume() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("getVolume() abgebrochen: Client-Key fehlt");
      return null;
    }
    const response = await this.callController(tv.address, tv.clientKey, "media.get_volume", null);
    if (!response) return null;
    if ((response as Record<string, unknown>)?.status === "error") {
      logger.warn({ response }, "getVolume() fehlgeschlagen");
      return null;
    }
    const volumeStatus = (response?.result as Record<string, unknown> | undefined)?.volumeStatus as
      | Record<string, unknown>
      | undefined;
    const volume = volumeStatus?.volume;
    return typeof volume === "number" ? volume : null;
  }

  static async setChannel(tv: LGTV, channelId: string) {
    logger.info({ address: tv.address }, "setChannel() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("setChannel() abgebrochen: Client-Key fehlt");
      return;
    }
    const escaped = channelId.replace(/\\/g, "\\\\").replace(/"/g, "\\\"");
    await this.callController(tv.address, tv.clientKey, "tv.set_channel_with_id", `"${escaped}"`);
  }

  static async startApp(tv: LGTV, appId: string) {
    logger.info({ address: tv.address }, "startApp() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("startApp() abgebrochen: Client-Key fehlt");
      return;
    }
    const escaped = appId.replace(/\\/g, "\\\\").replace(/"/g, "\\\"");
    await this.callController(tv.address, tv.clientKey, "application.launch", `"${escaped}"`);
  }

  static async notify(tv: LGTV, message: string) {
    logger.info({ address: tv.address }, "notify() fuer LGTV");
    if (!message) {
      logger.warn("notify() abgebrochen: Nachricht fehlt");
      return;
    }
    const escaped = message.replace(/\\/g, "\\\\").replace(/"/g, "\\\"");
    await this.callController(tv.address, tv.clientKey, "system.notify", `{"message":"${escaped}"}`);
  }

  static async getChannels(tv: LGTV) {
    logger.info({ address: tv.address }, "getChannels() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("getChannels() abgebrochen: Client-Key fehlt");
      return null;
    }
    const response = await this.callController(tv.address, tv.clientKey, "tv.channel_list", null);
    return parseChannels(response);
  }

  static async getApps(tv: LGTV) {
    logger.info({ address: tv.address }, "getApps() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("getApps() abgebrochen: Client-Key fehlt");
      return null;
    }
    const response = await this.callController(tv.address, tv.clientKey, "application.list_apps", null);
    return parseApps(response);
  }

  static async getSelectedApp(tv: LGTV) {
    logger.info({ address: tv.address }, "getSelectedApp() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("getSelectedApp() abgebrochen: Client-Key fehlt");
      return null;
    }
    const response = await this.callController(tv.address, tv.clientKey, "application.get_current", null);
    if (!response) return null;
    if ((response as Record<string, unknown>)?.status === "error") {
      logger.warn({ response }, "getSelectedApp() fehlgeschlagen");
      return null;
    }
    return typeof response?.result === "string" ? response?.result : null;
  }

  static async getSelectedChannel(tv: LGTV) {
    logger.info({ address: tv.address }, "getSelectedChannel() fuer LGTV");
    if (!tv.clientKey) {
      logger.warn("getSelectedChannel() abgebrochen: Client-Key fehlt");
      return null;
    }
    const response = await this.callController(tv.address, tv.clientKey, "tv.get_current_channel", null);
    if (!response) return null;
    if ((response as Record<string, unknown>)?.status === "error") {
      logger.warn({ response }, "getSelectedChannel() fehlgeschlagen");
      return null;
    }
    const result = response?.result as Record<string, unknown> | undefined;
    const channelId = result?.channelId;
    return typeof channelId === "string" ? channelId : null;
  }

  private static async callController(
    address: string | undefined,
    clientKey: string | null | undefined,
    event: string,
    params: string | null
  ): Promise<JsonValue> {
    if (!address) return null;
    if (event !== "wol" && !clientKey) return null;

    const scriptPath = resolveScriptPath("scripts", "pywebostv", "controller.py");
    const args = ["--ip", address];
    if (clientKey) {
      args.push("--client-key", clientKey);
    }
    args.push("--event", event);
    if (params != null) {
      args.push("--params", params);
    }
    const { output, code } = await runPython(scriptPath, args);
    if (code !== 0) {
      logger.error(
        { code, output: output.trim(), args },
        "PyWebOSTV-Skript beendet mit Exit-Code fuer {}",
        address
      );
    }
    const lastJsonLine = output
      .split(/\r?\n/)
      .map(line => line.trim())
      .filter(line => line.startsWith("{") && line.endsWith("}"))
      .pop();
    if (!lastJsonLine) return null;
    try {
      return JSON.parse(lastJsonLine) as JsonValue;
    } catch (err) {
      logger.warn({ err }, "PyWebOSTV-Output ist kein gültiges JSON: {}", lastJsonLine);
      return null;
    }
  }
}

function parseChannels(response: JsonValue) {
  if (!response) return null;
  if ((response as Record<string, unknown>)?.status === "error") {
    logger.warn({ response }, "getChannels() fehlgeschlagen");
    return null;
  }
  const result = response.result as Record<string, unknown> | undefined;
  const channelList = result?.channelList as Array<Record<string, unknown>> | undefined;
  if (!Array.isArray(channelList)) return null;
  return channelList
    .map(channelObj => {
      const id = channelObj.channelId as string | undefined;
      const channelNumber = typeof channelObj.majorNumber === "number" ? channelObj.majorNumber : undefined;
      const channelType = channelObj.channelMode as string | undefined;
      const hd = typeof channelObj.HDTV === "boolean" ? channelObj.HDTV : undefined;
      const name = channelObj.channelName as string | undefined;
      const imgUrl = channelObj.imgUrl as string | undefined;
      return new Channel(id, name, channelNumber, channelNumber, channelType, hd, imgUrl);
    })
    .filter(channel => Boolean(channel.id));
}

function parseApps(response: JsonValue) {
  if (!response) return null;
  if ((response as Record<string, unknown>)?.status === "error") {
    logger.warn({ response }, "getApps() fehlgeschlagen");
    return null;
  }
  const result = response.result as Array<Record<string, unknown>> | undefined;
  if (!Array.isArray(result)) return null;
  const apps: App[] = [];
  result.forEach((appObj, idx) => {
    const id = appObj.id as string | undefined;
    const name = appObj.title as string | undefined;
    const imgUrl = appObj.icon as string | undefined;
    if (!id || !name) return;
    apps.push(new App(id, name, imgUrl, idx + 1));
  });
  return apps;
}

async function runPython(scriptPath: string, args: string[]) {
  return new Promise<{ output: string; code: number | null }>(resolve => {
    const child = spawn("python", [scriptPath, ...args], { windowsHide: true });
    let output = "";
    child.stdout.on("data", chunk => {
      output += chunk.toString("utf8");
    });
    child.stderr.on("data", chunk => {
      output += chunk.toString("utf8");
    });
    child.on("close", code => {
      resolve({ output, code });
    });
  });
}

function escapeArgIfWindows(value: string) {
  return value;
}

function resolveScriptPath(...parts: string[]) {
  let current = process.cwd();
  for (let i = 0; i < 8; i += 1) {
    const candidate = path.resolve(current, ...parts);
    if (fs.existsSync(candidate)) {
      return candidate;
    }
    current = path.resolve(current, "..");
  }
  return path.resolve(process.cwd(), ...parts);
}

